### PR TITLE
Livestreaming typing animation

### LINKED ISSUE #4876: WebChat: Support for ‘stream’ output in ChatGPT style while using Azure OpenAI Completion
Bot responses with Azure OpenAI Completion, but we need the response to be displayed in a live streaming format.

### Describe the suggestion or request in detail

Allow Application that uses Bot framework webchat to stream out the response like ChatGPT web chat does.

### LINKED ISSUE #4939: SSR (Server Sent Event support)/Streaming support in WebChat
### Is your feature request related to a problem?

If we have to implement functionality like https://chat.openai.com/ where answer is rendered word (or few words) at a time instead of getting the complete answer in one shot, would it be possible to achieve it with BotFramework WebChat? If we get streamed answer from an LLM (say as SSR), which we can intercept and  and send "a", "a b", "a b c", "a b c d" and so on, on the WebSocket, is there some way in which we can make WebChat render it like how we see it in https://chat.openai.com/?

### Describe the suggestion or request in detail

_No response_

### Describe alternatives you have considered

_No response_

### Additional context

_No response_

### LINKED ISSUE #5184: Customize typing indicator duration via `channelData`
### Is your feature request related to a problem?

Some bots may take more than 5 seconds to finish a task.

It is not trivial to build a bot that send "typing" indicator every 5 seconds while the background task is ongoing.

### Describe the suggestion or request in detail

Will be great if there is a way to override `typingIndicatorDuration` on a per-activity basis, possibly in Web Chat-specific `channelData`.

### Describe alternatives you have considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by compulim**
> Fixes #4876. Fixes #4939. Fixes #5184.

### Added

-  Resolves [#5184](https://github.com/microsoft/BotFramework-WebChat/issues/5184). Added `channelData.webChat.styleOptions.typingIndicatorDuration` to override the default typing indicator duration on a per-activity basis, by [@compulim](https://github.com/compulim), in PR [#5141](https://github.com/microsoft/BotFramework-WebChat/pull/5141)
-  Resolves [#4876](https://github.com/microsoft/BotFramework-WebChat/issues/4876) and [#4939](https://github.com/microsoft/BotFramework-WebChat/issues/4939). Added support of livestreaming, by [@compulim](https://github.com/compulim), in PR [#5141](https://github.com/microsoft/BotFramework-WebChat/pull/5141)

## Description

Added livestreaming support and per-activity typing indicator duration.

Currently, we only support streaming contents. In future, we will support informative message indicating streaming status.

### Livestreaming

Add livestreaming support via "typing" activity with "text" field.

The following JSONs are not strictly enforced in Web Chat. However, for best compatibility across clients, bot authors are expected to follow the format.

Support of attachments and other features varies across clients. Bot authors should test their bot on individual clients.

Multiple simultaneous livestream session is supported. Make sure each session has its own unique ID.

#### First chunk

The first chunk indicates start of a livestream session.

```js
{
  channelData: {
    streamSequence: 1,      // Must be a monotonic increasing positive integer
    streamType: 'streaming' // Must be "streaming"
  },
  text: 'A quick',          // Content of the first chunk
  type: 'typing'            // Must be "typing"
}
```

#### Subsequent chunks

```js
{
  channelData: {
    streamId: '<activity ID of the first chunk>', // Must be activity ID of the first chunk
    streamSequence: 2,                            // Must be a monotonic increasing positive integer
    streamType: 'streaming'                       // Must be "streaming"
  },
  text: 'A quick brown fox',                      // Content must include past chunks
  type: 'typing'                                  // Must be "typing"
}
```

#### Final chunk

The final chunk will "close" the livestreaming session with a finalized message.

```js
{
  channelData: {
    streamId: '<activity ID of the first chunk>',       // Must be activity ID of the first chunk
    streamType: 'final'                                 // Must be "final"
  },
  text: 'A quick brown fox jumped over the lazy dogs.', // The whole message
  type: 'message'                                       // Must be "message"
}
```

### Per-activity typing indicator duration

Bot author can specify how long a typing indicator should show up for a specific activity via `channelData.webChat.styleOptions.typingIndicatorDuration`.

Currently, the `channelData.webChat.styleOptions` only supports `typingIndicatorDuration`. The broader support of `styleOptions` is being tracked by #5185.

## Specific Changes

- Added `useGetActivitiesByKey` to support "multiple activity revisions"
   - Livestream is more-or-less "multiple revisions of the same activity"
- Added `WebChatActivity.channelData.webChat.styleOptions.typingIndicatorDuration` property
   - When specified, will override the `styleOptions.typingIndicatorDuration` for the specific activity
- Added `WebChatActivity.channelData.webChat.receivedAt` property to indicate when the activity is received by Web Chat
- Updated `HOOKS.md` to include the `useGetActivitiesByKey` and some missing hooks

-  [x] I have added tests and executed them locally
-  [x] I have updated `CHANGELOG.md`
-  [x] I have updated documentation

## Review Checklist

> This section is for contributors to review your work.

-  [x] ~Accessibility reviewed (tab order, content readability, alt text, color contrast)~
-  [x] ~Browser and platform compatibilities reviewed~
-  [x] CSS styles reviewed (minimal rules, no `z-index`)
-  [x] Documents reviewed (docs, samples, live demo)
-  [x] ~Internationalization reviewed (strings, unit formatting)~
-  [x] ~`package.json` and `package-lock.json` reviewed~
-  [x] ~Security reviewed (no data URIs, check for nonce leak)~
-  [x] Tests reviewed (coverage, legitimacy)

### CODE REVIEW

File: __tests__/html/hooks.useActiveTyping.variable.html [hunk 1]
```diff
@@ -1,4 +1,4 @@
-<!DOCTYPE html>
+<!doctype html>
```
Comment (OEvgeny) [inline __tests__/html/hooks.useActiveTyping.variable.html:1]: :100:

Comment (compulim) [reply __tests__/html/hooks.useActiveTyping.variable.html:1]: Done by Prettier. 😉

**Review by tdurnford — APPROVED** [review]

File: packages/api/src/hooks/Composer.tsx [hunk 2]
```diff
@@ -589,14 +591,18 @@ const ComposerCore = ({
...
-          <SendBoxToolbarMiddlewareProvider middleware={sendBoxToolbarMiddleware || Object.freeze([])}>
-            {typeof children === 'function' ? children(context) : children}
-            <ActivitySendStatusTelemetryComposer />
-          </SendBoxToolbarMiddlewareProvider>
-        </SendBoxMiddlewareProvider>
-      </ActivitySendStatusComposer>
+      <ActivityListenerComposer>
+        <ActivitySendStatusComposer>
+          <ActivityTypingComposer>
+            <SendBoxMiddlewareProvider middleware={sendBoxMiddleware || Object.freeze([])}>
```
Comment (OEvgeny) [inline packages/api/src/hooks/Composer.tsx:597]: nit: probably a good idea to memorize the value

**Review by OEvgeny — APPROVED** [review]

File: packages/api/src/hooks/Composer.tsx (hunk 2 above)
Comment (compulim) [reply packages/api/src/hooks/Composer.tsx]: Fixed. 👍🏻

**Review by OEvgeny — APPROVED** [review]

**Comment by anshumankmr** [comment]
Awesome. Can't wait to t try this out👍👍👍

**Comment by rliberoff** [comment]
Hi @compulim,

Do you have an estimate when We can use this with the Bot Framework SDK?

**Comment by SomeDevFromSociabble** [comment]
> Hi @compulim,
> 
> Do you have an estimate when We can use this with the Bot Framework SDK?

I am interesting in testing it quickly too. 
The daily builds looked to be the solution but the last one was on february first on a commit of January 2020 (see #https://github.com/microsoft/BotFramework-WebChat/releases/daily).

**Comment by OEvgeny** [comment]
Releases from `main` available on npm via `@main` tag

**Pull request merged**